### PR TITLE
[Code Health] chore: define factory interfaces

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Description

Formalizes the convention we've established around module's `#Create()` factory methods. Additionally, defines a generalized form which has a generic return value, as well as variations for receiving a generic required arg, variadic optional arg(s), or both.

## Issue

Relates to #642.

Also came up during code review:
image

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [X] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Added `modules.ModuleFactoryWithOptions` interface
- Added factory interfaces:
  - `modules.FactoryWithRequired`
  - `modules.FactoryWithOptions`
  - `modules.FactoryWithRequiredAndOptions`
- Embedded `ModuleFactoryWithOptions` in `Module` interface

## Testing

- [ ] `make develop_test`; if any code changes were made
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/providers/peerstore_provider/peerstore_provider.go [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package peerstore_provider
 
-//go:generate mockgen -source=$GOFILE -destination=../../types/mocks/peerstore_provider_mock.go -package=mock_types github.com/pokt-network/pocket/p2p/types PeerstoreProvider
+//go:generate mockgen -destination=../../types/mocks/peerstore_provider_mock.go -package=mock_types github.com/pokt-network/pocket/p2p/providers/peerstore_provider PeerstoreProvider
```
Comment (bryanchriswhite) [inline p2p/providers/peerstore_provider/peerstore_provider.go:3]: ### Rational behind 89548d999ab1ce30cfc6df5d9be425ef95cc780e

@reviewer `mockgen` source mode seems to have trouble with embedded interfaces:

image

(see: https://github.com/golang/mock/issues/10)

Comment (Olshansk) [reply p2p/providers/peerstore_provider/peerstore_provider.go:3]: I think calling things out explicitly is even better

File: shared/modules/factory.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package modules
+
+// ModuleFactoryWithOptions implements a `#Create()` factory method which takes
+// a variadic `ModuleOption` argument(s) and returns a `Module`and an error.
+type ModuleFactoryWithOptions FactoryWithOptions[Module, ModuleOption]
+
+// FactoryWithRequired implements a `#Create()` factory method which takes a
+// required "config" argument of type K and returns a value of type T and an error.
+type FactoryWithRequired[T interface{}, K interface{}] interface {
```
Comment (Olshansk) [inline shared/modules/factory.go:9]: `FactoryWithConfig`?

File: shared/modules/factory.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// variadic "optional" argument(s) of type O and returns a value of type T
+// and an error.
+type FactoryWithOptions[T interface{}, O interface{}] interface {
+	Create(bus Bus, opts ...O) (T, error)
+}
+
+// FactoryWithRequiredAndOptions implements a `#Create()` factory method which
+// takes both a required "config" argument of type K and a variadic "optional"
+// argument(s) of type O and returns a value of type T and an error.
+type FactoryWithRequiredAndOptions[T interface{}, K interface{}, O interface{}] interface {
```
Comment (Olshansk) [inline shared/modules/factory.go:23]: `FactoryWithConfigAndOptions`?

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Review by Olshansk — APPROVED** [review]

**Pull request merged**